### ISSUE TITLE
Fix all plotting quirks by implementing a new graphics device

### ISSUE THREAD

**Issue by flying-sheep** [IRkernel/IRkernel#168]
using [RGraphicsDevice](http://www.omegahat.org/RGraphicsDevice/) we can define a device.

we could use this to record the primitive calls ourselves without the problems arising from e.g. using PDF (like warnings if the default PDF font doesn’t have some symbol).

it would probably also fix problem arising from things like “plot_builds_upon” and problems with margins and so on.

**Comment by takluyver** [comment]
I think something like that is what generates the recordedplot object already. I'm not confident we'd come up with something better than the mechanism already in evaluate.

**Comment by flying-sheep** [comment]
hmm. [here](http://www.omegahat.org/RGraphicsDevice/overview.html) are the functions one needs to implement. no idea how easy it is to handle things like `layout()` and `par(mfrow = ...)` gracefully.

but we’d have more control than `recordPlot`, e.g. we could define “one plot” in a better way than by our `plot_builds_upon` hackery. in reality, there’s plots and frames (i.e. layers), but recordplot doesn’t know about them

on the other hand i just saw that omegahat software has to be installed from their repo, so if we want to go CRAN one day we probably don’t want to depend on that.

**Comment by JanSchulz** [comment]
I'm not sure I see the problem: what would this fix?

I would still argue that improving `recordPlot`  would be a better longterm fix, as the knitr land would also benefit and they are currently a much heavier user of `recordPlot`

**Comment by takluyver** [comment]
`plot_builds_upon` is equivalent to some logic in knitr (though not a copy of the code, because GPL...). I know it's somewhat hackish, but it's not very big and hopefully shouldn't need to change very often, so I don't think there's a pressing need to replace it.

**Comment by flying-sheep** [comment]
@JanSchulz said:

> IMO: the current benchmark (and probably for the time being the much more used case) is knitr. All the problems we have are probably also present in knitr (e.g. the usecases for repr is more or less the same as for the asis objects knitr invented, just that they don't record the output format in the object and do not make it dependent on the runtime environment but a dev time decision. This is (/was) my argument why to talk to the knitr guys and propose to merge these two approaches). In this case I think we should only build a new plotting device if knitr also uses it, as otherwise the benefits are not outweighing the drawbacks of building and maintaining such a device.

**Comment by flying-sheep** [comment]
the limitations of our and knitrs hack are showing in all the unclosed plotting bugs. 190 wasn’t a bug, but all [unclosed label:plotting](https://github.com/IRkernel/IRkernel/labels/plotting) issues still are here.

 Duncan Murdoch explains in [R bug 16378](https://bugs.r-project.org/bugzilla/show_bug.cgi?id=16378#c1) that the plot quality suffers from our approach

also there is no way to do a crossplatform null device: `pdf(NULL)` throws warning when encountering unicode not available in the font (and maybe other stuff) and `png(NULL)` doesn’t work. `png('/dev/null')` is not portable.

**Comment by takluyver** [comment]
I'm not sure how many issues this would fix. We fixed a bunch of things by being smarter about which output formats we draw, and by ensuring SVG plots are isolated - those were problems in recreating the plots, not in capturing them. #190 is, I think, a misunderstanding rather than a problem. #186 looks like an error in creating a plot device to re-draw to, not in capturing. #123 looks like there's some specific knitr integration with ggvis, which a graphics device wouldn't give us.

That leaves #91 and #113 which *might* be fixed by creating a new graphics device. And it would doubtless introduce some bugs of its own.

I'm also inclined to agree with @JanSchulz that if this approach is good enough for knitr, which is far more popular than us (for now ;-), then it's good enough for us (...for now).

So, with my mini BDFL hat on: I don't want to do this under the IRkernel umbrella at the moment. We can reconsider if we get a lot more real-world reports of problems in plotting, and we're confident that writing a new graphics device will resolve them. Or if you can convince the knitr developers of the problem and the solution, I'd be happy to see a jointly maintained library that both projects adopt.

**Comment by flying-sheep** [comment]
reasonable! i guess i just don’t like hacks :stuck_out_tongue_winking_eye:

**Comment by flying-sheep** [comment]
OK. this shit needs to be done. pictex (see #341) *also* prints warnings in some cases.

at first we can simply use RGraphicsDevice to create a dummy device that does nothing but record stuff, and then maybe do smarter things with is. provided they put it on CRAN so we can actually use it (omegahat/RGraphicsDevice#1)

**Comment by janschulz** [comment]
Ok, but then, until we have a better solutions, we should simply go back to making a more generic one (aka png) the default and just use a tempfile? It will waste a bit of ram/diskspace, but it is mostly guaranteed to work...

**Comment by flying-sheep** [comment]
no, `png('/dev/null)` doesn’t work at all for OSX users, remember?

**Comment by janschulz** [comment]
I meant `png(tempfile())`.

**Comment by flying-sheep** [comment]
when will we delete this file then? is there a `dev.off()` hook?

**Comment by janschulz** [comment]
One could add a tempdir in the kernel startup and delete the complete dir in the end?

Or let the `null_device ` function write the tempfile to a global list and when the evaluate returns we delete all files in there?

**Comment by flying-sheep** [comment]
or we could continue to not create a file at all and avoid making things more complicated without a convincing reason.

**Comment by takluyver** [comment]
Maybe I'm just misreading what's normal conversation in Germany, but allow me to deploy a tactic we use when conversations get a bit heated...

Kittens!

![kitten](https://cloud.githubusercontent.com/assets/327925/15586225/27478d96-237b-11e6-9624-1dfe9b01571f.jpg)

**Comment by flying-sheep** [comment]
very sorry if this came over as if i was peeved at @janschulz!

the platform differences are a bit frustrating and the refusal of upstream R to consider re-adding a device like the one we’d need is inconvenient.

i think building a complex machinery to avoid the warnings on OSX isn’t worth it, and that i’ve already invested too much time into this arcane amassment of idiosyncrasies.

**Comment by flying-sheep** [comment]
New quirk that may make this worth considering: #658
